moveit_commander.MoveGroupInterface plan() to better align with C++ MoveGroup::plan()

bmagyar opened the pull request:
### Description

The `MoveGroupInterface.plan()` was out of line in its behaviour from the C++ class: now it returns a tuple of 4 as such: `(success flag : boolean, trajectory message : RobotTrajectory, planning time : float, error code : MoveitErrorCodes)`. 

Up to now, users of this function could only tell if a `plan()` call was successful by processing the trajectory message (some planners return trajectories regardless of a failed plan!).

It would be nice to add some tests too.

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [x] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

v4hn commented:
Only the last commit here breaks API (`plan` returning a tuple).
Could you please file a separate request for this commit and remove it here?
I would propose to merge this into L+ only.
Also I would like to propose to return a new class that can be cast to bool to simplify transition from the previous API.

Concerning this request: Could you please elaborate on how you interpret and use the `trajectory_constraints` field? I am not aware of any official documentation on what exactly it means (especially in contrast to `path_constraints`).

v4hn commented:
I updated my comment above to account for your changes.

bmagyar commented:
Didn't we just remove the bool casting from MoveItErrorCodes recently?
Casting to bool is a very C++ way of thinking, tuples are more Python-esque.

v4hn commented:
> Didn't we just remove the bool casting from MoveItErrorCodes recently?

We removed *implicit* bool casts. In that case we had to tell users to change 
`bool success = group.move()` to `bool success = static_cast<bool>(group.move())`.

>  Casting to bool is a very C++ way of thinking, tuples are more Python-esque.

With this PR we would have to tell them to change `group.plan()` to `group.plan()[0]` which I find rather unintuitive. I would believe `bool(group.plan())` or `group.plan().success` would be preferred over this.
That being said, I don't write a lot of Python code at the moment, so if you disagree we should wait for more comments from others..

bmagyar commented:
The current Python API only returns a `moveit_msgs.msg.RobotTrajectory` object which doesn't have bool casting options either. The way `deserialize()` runs on a Python ROS message object there is no possibility for that to become a `None` object which removes boolean conversion also for the edge-case of a bad message.

rhaschke commented:
@bmagyar Please fix clang-format issues.

bmagyar commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -471,9 +472,16 @@ def plan(self, joints = None):
         plan = RobotTrajectory()
-        plan.deserialize(self._g.compute_plan())
-        return plan
+        return (error_code.val == MoveItErrorCodes.SUCCESS,
```
Done!

rhaschke replied:
While I like the idea to match the C++ interface, this change essentially changes API of the `plan()` function in Python.
Do we want to accept this for Melodic as well? Or should this be limited to the master branch? Please also add a note to `MIGRATION.md`.

bmagyar replied:
I received the very same comment from a year ago where it was proposed to only add it to Lunar, not Kinetic.

I think it's ok to put this on melodic too.

bmagyar commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -471,9 +472,16 @@ def plan(self, joints = None):
+        error_code = MoveItErrorCodes()
+        error_code.deserialize(error_code_msg)
         plan = RobotTrajectory()
-        plan.deserialize(self._g.compute_plan())
```
I've changed it from to `plan` to reflect the `MoveGroupInterface` better.
```
MoveItErrorCode plan(Plan& plan)
```

I can remove `compute_path`

rhaschke replied:
Please do so. Am I correct, that these names are internal only?

bmagyar replied:
So internal, some of them - include `compute_plan()` only exist as a string in the wrapping rules.

bmagyar commented:
Sorry about the rebase, got a conflict with #788 I had to resolve.

rhaschke commented:
We have failing tests now. Is that due to the rebase?

bmagyar commented:
What a pleasant surprise! 
It comes from the tests is related to the removed `compute_plan()` function. I'll update them now.

rhaschke commented:
> It comes from the tests is related to the removed compute_plan() function. I'll update them now.

Hm. If `compute_plan()` was exposed and intended for use, we should only deprecate it for now (and pointing to the implementation of `plan()`). That's why I asked for it being _internal_ only.

bmagyar commented:
It is _internal only_. This [test](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning_interface/test/python_move_group.py) may look familiar to you :) 

Implementing a unit test of a C++ class through it's Python-wrapper is risky. For instance, you can get all kinds of issues such as this one :)

bmagyar commented:
I've added checking on the error codes too. Unfortunately it's a bit ugly now since these tests are using the raw C++ to Python interface I had to use `deserialize` on the returned values to get a `MoveItErrorCode` object. Let me know what you think :)

rhaschke commented:
Feel free to modify the test to use the python-only interface. Obviously I was not aware of this interface, when writing the tests...

v4hn commented:
> The current Python API only returns a moveit_msgs.msg.RobotTrajectory object which doesn't have bool casting options either

> in python, where you simply write
`err, trajectory, time = plan()`

Fair enough. It's actually `err, trajectory, time, errcode = plan()` in the proposed code.
In my opinion that looks like someone really wants matlab's variable-returns support in Python..

I take back my comments though :+1:.

bmagyar commented:
Yes! I am planning to do that next weekend.

henningkayser commented:
@bmagyar Do you think you could finish this PR and target it to the new master branch?

bmagyar commented:
@rhaschke yes I can bundle that into this one.
@henningkayser Sure no problem.

I hope it's ok if I only get to this on the weekend

bmagyar commented:
@rhaschke I am not getting test failures locally with these changes. I went back to see what we discussed exactly about those tests and I have to say that I think I've changed my mind since then. I think there is justification for any sort of coverage no matter which side of a binding it is. 

There is a TODO issue I'd like to add to follow this PR up. We should merge the two Python test implementations, `python_move_group_ns.py` and `python_move_group.py` as they only differ in a single parameter. It could be parametrized using 2 `.test` files passing in different arguments to the `.py` file or adding a 3-rd file which contains a parametrized Python class, the two individual test nodes instantiating it with different parameter values. This simple refactoring would remove 100 lines of code and ease maintaining.

rhaschke commented:
> I went back to see what we discussed ...

I'm lost. Could you please add links for the mentioned "back" reference(s).

> There is a TODO issue.

Could you also provide a (source code) link to this TODO. Or is the paragraph in https://github.com/ros-planning/moveit/pull/790#issuecomment-471365983 describing the todo, i.e. merging `python_move_group_ns.py` and `python_move_group.py`? A big :+1:  for this idea!

bmagyar commented:
> I'm lost. Could you please add links for the mentioned "back" reference(s).

Sorry for the brevity of that. I was referring to [this comments](https://github.com/ros-planning/moveit/pull/790#issuecomment-372312358), more specifically to [this test](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning_interface/test/python_move_group.py). While using `deserialize` may not be the nicest, I think any means of providing coverage for code is a good means, and it can only get better from there :) 

> Could you also provide a (source code) link to this TODO. Or is the paragraph in #790 (comment) describing the todo, i.e. merging python_move_group_ns.py and python_move_group.py? A big +1 for this idea!

Yes, I'm not a big fan of in-code todo-s in general, we have an issue tracker for that :)

bmagyar commented:
@rhaschke I think your comments may be resolved now.
@mlautman any comments on this?

rhaschke commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -316,10 +317,6 @@ def set_random_target(self):
         """ Set a random joint configuration target """
         self._g.set_random_target()
 
-    def get_named_targets(self):
```
I don't see a reason to remove this. I think it's by mistake.

bmagyar replied:
Correct, I had added it back (in a new commit, the MR should be squashed anyways

rhaschke commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -425,11 +425,17 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
   }
 
-  std::string getPlanPython()
+  const char* getPlannerIdCStr() const
```
```suggestion
  const std::string& getPlannerId() const
```

bmagyar commented on moveit_ros/planning_interface/test/robot_state_update.py:
```diff
@@ -22,14 +22,17 @@ def tearDown(self):
+            error_code1, plan, time = self.plan(target)
+            error_code = MoveItErrorCodes()
+            error_code.deserialize(error_code1)
+            if self.group.execute(plan) and (error_code.val == MoveItErrorCodes.SUCCESS):
```
:+1:

bmagyar commented on moveit_ros/planning_interface/test/python_move_group.py:
```diff
@@ -53,8 +62,11 @@ def test_validation(self):
         # newly planned trajectory should execute again
-        plan3 = self.plan(current)
+        error_code3, plan3, time = self.plan(current)
         self.assertTrue(self.group.execute(plan3))
```
done

rhaschke commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -425,9 +425,9 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
     return asyncExecute(plan) == MoveItErrorCode::SUCCESS;
   }
 
-  const char* getPlannerIdCStr() const
```
Oops. This became an infinite recursive function call now. Where is this function (or the previous `getPlannerIdCStr`) used at all?

bmagyar replied:
I think we tried to fix something that git already did, or perhaps it was on an outdated diff. This conversation seems to be on an outdated diff as well... :/ 

The current diff shows me adding this function which is absolutely not intended. Big of a jump from a PR intended for Kinetic a year ago :)

bmagyar replied:
The github diff view seems to be cached and I discovered a refresh button on the top... That may be the trick. I've added a new commit to remove the function, compilation seems fine, let's see travis with the tests.

henningkayser commented:
@rhaschke good to go?

davetcoleman commented:
+1 to not changing melodic beyond bug fixes.

bmagyar commented:
That makes it 3 vs 1 for not going to Melodic, it's decided then :) 

I've rebased the branch, dropped the migration notes for Melodic and pre-squashed the commits, all ready for merging as soon as the CI turns green.

rhaschke commented:
@bmagyar, please re-add the MIGRATION notes, but below section Noetic.

bmagyar commented:
updated

rhaschke commented on MIGRATION.md:
```diff
@@ -3,6 +3,7 @@
 API changes in MoveIt! releases
 
 ## ROS Noetic (upcoming changes in master)
+- Extended the return of `MoveitCommander.MoveGroup.plan()` from `trajectory` to a tuple of `(success, trajectory, planning_time, error_code)` to better match the C++ interface of MoveGroupInterface ([790](https://github.com/ros-planning/moveit/pull/790/))
```
```suggestion
- Extended the return value of `MoveitCommander.MoveGroup.plan()` from `trajectory` to a tuple of `(success, trajectory, planning_time, error_code)` to better match the C++ MoveGroupInterface ([790](https://github.com/ros-planning/moveit/pull/790/))
```

bmagyar commented:
That took a while, thanks for sticking around! :muscle:
